src/bgg/circuit/mod.rs

```diff
@@ -233,75 +208,120 @@ impl PolyCircuit {
-                    let right =
-                        wires.get(&input_ids[1]).expect("wire value missing for Add").clone();
-                    wires.insert(gate_id, left + right);
+                    let left = wires[gate.input_gates[0]].as_ref().expect("wire missing for Add");
+                    let right = wires[gate.input_gates[1]].as_ref().expect("wire missing for Add");
+                    left.clone() + right
                 }
                 PolyGateType::Sub => {
-                    let left =
-                        wires.get(&input_ids[0]).expect("wire value missing for Sub").clone();
-                    let right =
-                        wires.get(&input_ids[1]).expect("wire value missing for Sub").clone();
-                    wires.insert(gate_id, left - right);
+                    let left = wires[gate.input_gates[0]].as_ref().expect("wire missing for Sub");
+                    let right = wires[gate.input_gates[1]].as_ref().expect("wire missing for Sub");
+                    left.clone() - right
                 }
                 PolyGateType::Mul => {
-                    let left =
-                        wires.get(&input_ids[0]).expect("wire value missing for Mul").clone();
-                    let right =
-                        wires.get(&input_ids[1]).expect("wire value missing for Mul").clone();
-                    wires.insert(gate_id, left * right);
+                    let left = wires[gate.input_gates[0]].as_ref().expect("wire missing for Mul");
+                    let right = wires[gate.input_gates[1]].as_ref().expect("wire missing for Mul");
+                    left.clone() * right
                 }
                 PolyGateType::Rotate { shift } => {
-                    let input = wires.get(&input_ids[0]).expect("wire value missing for Rotate");
-                    wires.insert(gate_id, input.rotate(params, *shift));
+                    let input =
+                        wires[gate.input_gates[0]].as_ref().expect("wire missing for Rotate");
+                    input.rotate(params, *shift)
                 }
-                PolyGateType::Call { circuit_id, output_id, .. } => {
-                    let sub_circuit = &self.sub_circuits[circuit_id];
-                    let sub_inputs: Vec<E> = input_ids
-                        .iter()
-                        .map(|id| wires.get(id).expect("wire value missing for Call").clone())
-                        .collect();
-                    let outputs = sub_circuit.eval(params, one, &sub_inputs);
-                    let first_output_wire = gate_id - output_id;
-                    for (i, out_val) in outputs.into_iter().enumerate() {
-                        wires.insert(first_output_wire + i, out_val);
-                    }
+                PolyGateType::Call { .. } => {
+                    panic!("no more call gate type during evaluation");
```

Comment:
@rkdud007 Can you just remove `Call` from `PolyGateType` if we no longer use it?

Reply:
We use this for register it, Call type got alter when call_sub_circuit method called :)